File: src/utils/jwtValidator.js
const jwt = require('jsonwebtoken');
const { logInfo, logError } = require('../logger');
const { config } = require('../config');

/**
 * Validates a JWT token
 * @param {string} token - The JWT token to validate
 * @returns {Promise<Object|null>} - The decoded payload if valid, null if invalid
 */
async function validateJWT(token) {
  try {
    if (!token) {
      logError('JWT validation failed: No token provided');
      return null;
    }

    // Remove 'Bearer ' prefix if present
    const cleanToken = token.replace(/^Bearer\s+/i, '');

    if (!config.security.jwtSecret) {
      logError('JWT validation failed: No JWT secret configured');
      return null;
    }

    const decoded = jwt.verify(cleanToken, config.security.jwtSecret);
    logInfo('JWT validation successful', { userId: decoded.userId, exp: decoded.exp });
    
    return decoded;
  } catch (error) {
    if (error.name === 'TokenExpiredError') {
      logError('JWT validation failed: Token expired', { exp: error.expiredAt });
    } else if (error.name === 'JsonWebTokenError') {
      logError('JWT validation failed: Invalid token', { message: error.message });
    } else if (error.name === 'NotBeforeError') {
      logError('JWT validation failed: Token not active', { notBefore: error.date });
    } else {
      logError('JWT validation failed: Unknown error', error);
    }
    return null;
  }
}

/**
 * Generates a JWT token
 * @param {Object} payload - The payload to encode
 * @param {string} expiresIn - Token expiration time
 * @returns {string} - The generated JWT token
 */
function generateJWT(payload, expiresIn = config.security.jwtExpiration) {
  try {
    if (!config.security.jwtSecret) {
      throw new Error('JWT secret not configured');
    }

    const token = jwt.sign(payload, config.security.jwtSecret, { expiresIn });
    logInfo('JWT generated successfully', { userId: payload.userId, expiresIn });
    
    return token;
  } catch (error) {
    logError('JWT generation failed', error);
    throw error;
  }
}

/**
 * Extracts JWT token from request headers
 * @param {Object} headers - Request headers
 * @returns {string|null} - The extracted token or null
 */
function extractTokenFromHeaders(headers) {
  const authorization = headers.authorization || headers.Authorization;
  
  if (!authorization) {
    return null;
  }

  // Handle both 'Bearer token' and just 'token' formats
  if (authorization.startsWith('Bearer ')) {
    return authorization.substring(7);
  }
  
  return authorization;
}

/**
 * Middleware function for validating JWT in Express-like frameworks
 * @param {Object} req - Request object
 * @param {Object} res - Response object
 * @param {Function} next - Next middleware function
 */
async function jwtMiddleware(req, res, next) {
  try {
    const token = extractTokenFromHeaders(req.headers);
    
    if (!token) {
      return res.status(401).json({ error: 'No token provided' });
    }

    const decoded = await validateJWT(token);
    
    if (!decoded) {
      return res.status(401).json({ error: 'Invalid token' });
    }

    req.user = decoded;
    next();
  } catch (error) {
    logError('JWT middleware error', error);
    res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  validateJWT,
  generateJWT,
  extractTokenFromHeaders,
  jwtMiddleware
};

File: python_app/utils/jwt_validator.py
from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from typing import Any, Callable, Dict, Mapping, Optional

import jwt

from ..config import config
from ..logger import log_error, log_info


def _parse_duration(value: str | None) -> Optional[int]:
    if not value:
        return None
    if value.isdigit():
        return int(value)
    match = re.match(r"^(\d+)([smhd])$", value.strip(), flags=re.IGNORECASE)
    if not match:
        return None
    amount = int(match.group(1))
    unit = match.group(2).lower()
    multiplier = {"s": 1, "m": 60, "h": 3600, "d": 86400}[unit]
    return amount * multiplier


def validate_jwt(token: str | None) -> Optional[Dict[str, Any]]:
    try:
        if not token:
            log_error("JWT validation failed: No token provided")
            return None

        clean_token = re.sub(r"^Bearer\s+", "", token, flags=re.IGNORECASE)

        if not config.security.jwt_secret:
            log_error("JWT validation failed: No JWT secret configured")
            return None

        decoded = jwt.decode(
            clean_token,
            config.security.jwt_secret,
            algorithms=["HS256"],
        )
        log_info(
            "JWT validation successful",
            {"userId": decoded.get("userId"), "exp": decoded.get("exp")},
        )
        return decoded
    except jwt.ExpiredSignatureError as error:
        log_error("JWT validation failed: Token expired", error)
    except jwt.InvalidTokenError as error:
        log_error("JWT validation failed: Invalid token", error)
    except Exception as error:  # pylint: disable=broad-except
        log_error("JWT validation failed: Unknown error", error)
    return None


def generate_jwt(
    payload: Mapping[str, Any],
    expires_in: Optional[str] = None,
) -> str:
    if not config.security.jwt_secret:
        raise ValueError("JWT secret not configured")

    payload_dict = dict(payload)
    duration = _parse_duration(expires_in or config.security.jwt_expiration)
    if duration:
        payload_dict["exp"] = datetime.now(timezone.utc) + timedelta(seconds=duration)

    token = jwt.encode(payload_dict, config.security.jwt_secret, algorithm="HS256")
    log_info(
        "JWT generated successfully",
        {"userId": payload_dict.get("userId"), "expiresIn": duration},
    )
    return token


def extract_token_from_headers(headers: Mapping[str, Any]) -> Optional[str]:
    authorization = headers.get("authorization") or headers.get("Authorization")
    if not authorization:
        return None
    if authorization.startswith("Bearer "):
        return authorization[7:]
    return authorization


def jwt_middleware(
    request: Any,
    respond: Callable[[int, Dict[str, Any]], Any],
    next_func: Callable[[], Any],
) -> Any:
    try:
        token = extract_token_from_headers(getattr(request, "headers", {}))
        if not token:
            return respond(401, {"error": "No token provided"})

        decoded = validate_jwt(token)
        if not decoded:
            return respond(401, {"error": "Invalid token"})

        setattr(request, "user", decoded)
        return next_func()
    except Exception as error:  # pylint: disable=broad-except
        log_error("JWT middleware error", error)
        return respond(500, {"error": "Internal server error"})
